File: typescript/src/delta.ts
import { open, rename, unlink } from "node:fs/promises";
import { dirname } from "node:path";

import { stringify as stringifyYaml } from "yaml";

import { profileDigest } from "./canonical.js";
import type { AgentProfile, AgentStateDelta, DeltaApplication, JsonObject, JsonValue } from "./types.js";

export class ApplyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ApplyError";
  }
}

export class ConflictError extends ApplyError {
  constructor(message: string) {
    super(message);
    this.name = "ConflictError";
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function pointerTokens(document: JsonObject, pointer: string): string[] {
  if (!pointer.startsWith("/")) throw new ApplyError(`pointer must start with '/': ${pointer}`);
  const raw = pointer.slice(1).split("/").map((token) => token.replaceAll("~1", "/").replaceAll("~0", "~"));
  const resolved: string[] = [];
  let cursor: unknown = document;
  for (const token of raw) {
    if (Array.isArray(cursor) && token.startsWith("id:")) {
      const id = token.slice(3);
      const index = cursor.findIndex((item) => isObject(item) && item.id === id);
      if (index < 0) resolved.push("-missing-");
      else resolved.push(String(index));
      cursor = index < 0 ? undefined : cursor[index];
      continue;
    }
    resolved.push(token);
    if (token === "-") cursor = undefined;
    else if (Array.isArray(cursor) && /^\d+$/.test(token)) cursor = cursor[Number(token)];
    else if (isObject(cursor)) cursor = cursor[token];
    else cursor = undefined;
  }
  return resolved;
}

function applyOperation(document: JsonObject, operation: JsonObject, warnings: string[]): void {
  const kind = operation.op;
  const pointer = operation.path;
  if (typeof pointer !== "string" || !(pointer === "/state" || pointer.startsWith("/state/"))) throw new ApplyError(`operation path ${JSON.stringify(pointer)} is outside /state`);
  if (kind !== "add" && kind !== "replace" && kind !== "remove") throw new ApplyError(`unknown operation ${JSON.stringify(kind)}`);
  const tokens = pointerTokens(document, pointer);
  if (tokens.includes("-missing-")) {
    if (kind === "remove") {
      warnings.push(`remove on missing path ${JSON.stringify(pointer)} ignored`);
      return;
    }
    throw new ApplyError(`path ${JSON.stringify(pointer)} does not resolve`);
  }
  let parent: unknown = document;
  for (const token of tokens.slice(0, -1)) {
    if (Array.isArray(parent) && /^\d+$/.test(token)) parent = parent[Number(token)];
    else if (isObject(parent)) parent = (parent[token] ??= {});
    else throw new ApplyError(`path ${JSON.stringify(pointer)} does not resolve`);
  }
  const last = tokens.at(-1);
  if (last === undefined) throw new ApplyError("operation path must not be document root");
  if (Array.isArray(parent)) {
    if (kind === "add") {
      if (last === "-") parent.push(structuredClone(operation.value));
      else if (/^\d+$/.test(last)) parent.splice(Number(last), 0, structuredClone(operation.value));
      else throw new ApplyError(`array index expected at ${pointer}`);
    } else if (!/^\d+$/.test(last) || Number(last) >= parent.length) {
      if (kind === "remove") warnings.push(`remove on missing path ${JSON.stringify(pointer)} ignored`);
      else throw new ApplyError(`index out of range at ${pointer}`);
    } else if (kind === "replace") parent[Number(last)] = structuredClone(operation.value);
    else parent.splice(Number(last), 1);
  } else if (isObject(parent)) {
    if (kind === "remove") {
      if (!(last in parent)) warnings.push(`remove on missing path ${JSON.stringify(pointer)} ignored`);
      else delete parent[last];
    } else parent[last] = structuredClone(operation.value);
  } else throw new ApplyError(`path ${JSON.stringify(pointer)} does not resolve to a container`);
}

function entrySortKey(entry: Record<string, unknown>, strategy: string): string | number {
  if (strategy === "least_confident") return typeof entry.confidence === "number" ? entry.confidence : 1;
  if (strategy === "oldest") return String(entry.learned_at ?? entry.opened_at ?? "");
  return String(entry.last_used_at ?? entry.updated_at ?? entry.learned_at ?? "");
}

function compareEntries(a: Record<string, unknown>, b: Record<string, unknown>, strategy: string): number {
  const left = entrySortKey(a, strategy);
  const right = entrySortKey(b, strategy);
  if (typeof left === "number" && typeof right === "number") return left - right;
  return String(left).localeCompare(String(right));
}

function enforceRetention(profile: AgentProfile, warnings: string[], currentTime: string): void {
  const spec = profile.spec;
  const lifecycle = isObject(spec.lifecycle) ? spec.lifecycle : {};
  const retention = isObject(lifecycle.retention) ? lifecycle.retention : {};
  const state = isObject(profile.state) ? profile.state : {};
  const strategy = typeof retention.eviction === "string" ? retention.eviction : "least_recently_used";
  for (const collection of ["facts", "preferences"]) {
    const rawEntries = state[collection];
    const entries: Array<Record<string, unknown>> | undefined = Array.isArray(rawEntries)
      ? (rawEntries as unknown[]).filter(isObject)
      : undefined;
    if (!entries) continue;
    let retained = entries;
    if (retention.fact_ttl_days) {
      retained = entries.filter((entry) => {
        const expired = typeof entry.expires_at === "string" && entry.expires_at < currentTime && entry.pinned !== true;
        if (expired) warnings.push(`evicted expired ${collection} entry ${JSON.stringify(entry.id)}`);
        return !expired;
      });
      state[collection] = retained as JsonValue;
    }
    const cap = collection === "facts" && typeof retention.max_facts === "number" ? retention.max_facts : undefined;
    if (cap === undefined || retained.length <= cap) continue;
    const pinned = retained.filter((entry) => entry.pinned === true);
    const rest = retained.filter((entry) => entry.pinned !== true).sort((a, b) => compareEntries(a, b, strategy));
    const room = Math.max(cap - pinned.length, 0);
    const evicted = rest.slice(0, Math.max(rest.length - room, 0));
    for (const entry of evicted) warnings.push(`evicted ${collection} entry ${JSON.stringify(entry.id)} (${strategy})`);
    const kept = new Set([...pinned, ...rest.slice(evicted.length)]);
    state[collection] = retained.filter((entry) => kept.has(entry)) as JsonValue;
  }
  const threadCap = typeof retention.max_open_threads === "number" ? retention.max_open_threads : undefined;
  const rawThreads = state.open_threads;
  const threads: Array<Record<string, unknown>> | undefined = Array.isArray(rawThreads)
    ? (rawThreads as unknown[]).filter(isObject)
    : undefined;
  if (threads && threadCap !== undefined && threads.length > threadCap) {
    const active = threads.filter((thread) => !thread.status || thread.status === "open" || thread.status === "blocked");
    const closed = threads
      .filter((thread) => thread.status === "done" || thread.status === "abandoned")
      .sort((a, b) => String(a.updated_at ?? "").localeCompare(String(b.updated_at ?? "")));
    const dropped = new Set(closed.slice(0, threads.length - threadCap));
    for (const thread of dropped) warnings.push(`evicted closed thread ${JSON.stringify(thread.id)}`);
    const remaining = threads.filter((thread) => !dropped.has(thread));
    state.open_threads = (remaining.length > threadCap ? remaining.slice(-threadCap) : remaining) as JsonValue;
    if (remaining.length > threadCap) warnings.push(`open_threads still over cap after evicting closed threads (${active.length} active)`);
  }
  const historyCap = typeof retention.max_history === "number" ? retention.max_history : 50;
  if (Array.isArray(profile.history) && profile.history.length > historyCap) profile.history = profile.history.slice(-historyCap);
}

export interface ApplyOptions {
  approved?: boolean;
  actor?: string;
  now?: () => string;
}

export function applyDelta(profile: AgentProfile, delta: AgentStateDelta, options: ApplyOptions = {}): DeltaApplication {
  const metadata = profile.metadata;
  const current = typeof metadata.revision === "number" ? metadata.revision : 1;
  if (delta.target.name !== metadata.name) throw new ApplyError(`delta targets ${JSON.stringify(delta.target.name)} but profile is ${JSON.stringify(metadata.name)}`);
  if (delta.target.revision !== current) throw new ConflictError(`delta targets revision ${delta.target.revision} but profile is at ${current}`);
  if (typeof delta.target.digest === "string" && delta.target.digest !== profileDigest(profile)) throw new ApplyError("target.digest does not match profile");
  const lifecycle = isObject(profile.spec.lifecycle) ? profile.spec.lifecycle : {};
  const writeback = typeof lifecycle.writeback === "string" ? lifecycle.writeback : "propose";
  if (writeback === "off") throw new ApplyError("lifecycle.writeback is 'off'");
  if (writeback === "propose" && !options.approved) throw new ApplyError("lifecycle.writeback is 'propose'; explicit approval is required");

  const working = structuredClone(profile);
  working.state ??= {};
  const warnings: string[] = [];
  const stamp = options.now?.() ?? new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
  const operations = Array.isArray(delta.operations) ? delta.operations.filter((value): value is JsonObject => isObject(value)) : [];
  operations.forEach((operation, index) => {
    try {
      applyOperation(working, operation, warnings);
    } catch (error) {
      throw new ApplyError(`operation ${index}: ${(error as Error).message}`);
    }
  });
  enforceRetention(working, warnings, stamp);
  working.metadata.revision = current + 1;
  working.metadata.updated_at = stamp;
  if (operations.length > 0 && isObject(working.state)) {
    working.state.updated_at = stamp;
    working.state.revision = (typeof working.state.revision === "number" ? working.state.revision : 0) + 1;
  }
  const session = delta.session;
  const history: JsonObject = {
    revision: current + 1,
    at: stamp,
    by: typeof session.id === "string" ? session.id : (options.actor ?? "oap-typescript"),
    change: typeof delta.summary === "string" ? delta.summary : `${operations.length} state operations`,
    sections: ["state"],
  };
  if (typeof session.id === "string") history.session_id = session.id;
  if (typeof session.harness === "string") history.harness = session.harness;
  if (options.approved) history.approved_by = options.actor ?? "oap-typescript";
  working.history = [...(Array.isArray(working.history) ? working.history : []), history];
  enforceRetention(working, warnings, stamp);
  const pendingProposals = (Array.isArray(delta.proposals) ? delta.proposals.filter((value): value is JsonObject => isObject(value)) : []).map((proposal) => {
    const copy = structuredClone(proposal);
    if (typeof copy.path === "string" && ["/spec/tools", "/spec/permissions", "/spec/memory", "/spec/runtime/subagents"].some((prefix) => String(copy.path).startsWith(prefix))) copy.risk = "high";
    return copy;
  });
  return { profile: working, warnings, pendingProposals };
}

export function serializeOap(document: JsonObject, format: "yaml" | "json" | "markdown" = "yaml"): string {
  if (format === "json") return `${JSON.stringify(document, null, 2)}\n`;
  if (format === "yaml") return stringifyYaml(document, { lineWidth: 100 });
  const copy = structuredClone(document);
  const spec = copy.spec as JsonObject;
  const role = spec.role as JsonObject;
  const instructions = String(role.instructions ?? "").trimEnd();
  delete role.instructions;
  return `---\n${stringifyYaml(copy, { lineWidth: 100 }).trimEnd()}\n---\n${instructions}\n`;
}

export async function writeAtomically(path: string, text: string): Promise<void> {
  const temporary = `${path}.${process.pid}.${Date.now()}.tmp`;
  let handle;
  try {
    handle = await open(temporary, "wx", 0o600);
    await handle.writeFile(text, "utf8");
    await handle.sync();
    await handle.close();
    handle = undefined;
    await rename(temporary, path);
    const directory = await open(dirname(path), "r");
    try {
      await directory.sync();
    } finally {
      await directory.close();
    }
  } catch (error) {
    if (handle) await handle.close().catch(() => undefined);
    await unlink(temporary).catch(() => undefined);
    throw error;
  }
}

File: rust/src/delta.rs
use std::{
    cmp::Ordering,
    fs::{self, OpenOptions},
    io::Write,
    path::Path,
};

use chrono::{DateTime, SecondsFormat, Utc};
use serde_json::{Map, Value, json};
use thiserror::Error;

use crate::{AgentProfile, AgentStateDelta, DeltaApplication, Document, object, profile_digest};

/// Failure while validating or applying an agent state delta.
#[derive(Debug, Error)]
pub enum ApplyError {
    /// Optimistic-concurrency target data did not match the profile.
    #[error(transparent)]
    Conflict(#[from] ConflictError),
    /// The delta was invalid or could not be applied or persisted.
    #[error("{0}")]
    Message(String),
}

/// Optimistic-concurrency conflict raised before any mutation is committed.
#[derive(Debug, Error)]
#[error("{0}")]
pub struct ConflictError(pub String);

/// Controls approval, attribution, and time for delta application.
#[derive(Default)]
pub struct ApplyOptions<'a> {
    /// Whether a `propose` writeback policy has explicit human approval.
    pub approved: bool,
    /// Actor recorded in the appended history entry.
    pub actor: Option<&'a str>,
    /// Deterministic timestamp override; current UTC time is used when absent.
    pub now: Option<DateTime<Utc>>,
}

fn pointer_tokens(pointer: &str) -> Vec<String> {
    pointer
        .trim_start_matches('/')
        .split('/')
        .filter(|part| !part.is_empty())
        .map(|part| part.replace("~1", "/").replace("~0", "~"))
        .collect()
}

fn modify(
    current: &mut Value,
    tokens: &[String],
    kind: &str,
    value: Option<&Value>,
) -> Result<bool, ApplyError> {
    if tokens.is_empty() {
        *current = value.cloned().unwrap_or(Value::Null);
        return Ok(false);
    }
    let token = &tokens[0];
    let last = tokens.len() == 1;
    match current {
        Value::Object(map) => {
            if last {
                match kind {
                    "remove" => return Ok(map.remove(token).is_none()),
                    "add" | "replace" => {
                        map.insert(token.clone(), value.cloned().unwrap_or(Value::Null));
                        return Ok(false);
                    }
                    _ => return Err(ApplyError::Message(format!("unknown operation {kind:?}"))),
                }
            }
            if !map.contains_key(token) {
                map.insert(token.clone(), Value::Object(Map::new()));
            }
            modify(map.get_mut(token).unwrap(), &tokens[1..], kind, value)
        }
        Value::Array(items) => {
            let index = if let Some(id) = token.strip_prefix("id:") {
                items.iter().position(|item| {
                    object(Some(item)).get("id").and_then(Value::as_str) == Some(id)
                })
            } else if token == "-" {
                Some(items.len())
            } else {
                token.parse::<usize>().ok()
            };
            let Some(index) = index else {
                return Ok(true);
            };
            if last {
                match kind {
                    "add" if index <= items.len() => {
                        items.insert(index, value.cloned().unwrap_or(Value::Null));
                        Ok(false)
                    }
                    "remove" if index < items.len() => {
                        items.remove(index);
                        Ok(false)
                    }
                    "replace" if index < items.len() => {
                        items[index] = value.cloned().unwrap_or(Value::Null);
                        Ok(false)
                    }
                    "add" | "remove" | "replace" => Ok(true),
                    _ => Err(ApplyError::Message(format!("unknown operation {kind:?}"))),
                }
            } else if index < items.len() {
                modify(&mut items[index], &tokens[1..], kind, value)
            } else {
                Ok(true)
            }
        }
        _ => Ok(true),
    }
}

fn apply_operation(
    document: &mut Document,
    operation: &Document,
    warnings: &mut Vec<String>,
) -> Result<(), ApplyError> {
    let kind = operation
        .get("op")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let pointer = operation
        .get("path")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if pointer != "/state" && !pointer.starts_with("/state/") {
        return Err(ApplyError::Message(format!(
            "operation path {pointer:?} is outside /state"
        )));
    }
    let mut root = Value::Object(std::mem::take(document));
    let missing = modify(
        &mut root,
        &pointer_tokens(pointer),
        kind,
        operation.get("value"),
    )?;
    *document = root.as_object().cloned().unwrap_or_default();
    if missing {
        if kind == "remove" {
            warnings.push(format!("remove on missing path {pointer:?} ignored"));
            Ok(())
        } else {
            Err(ApplyError::Message(format!(
                "path {pointer:?} does not resolve"
            )))
        }
    } else {
        Ok(())
    }
}

/// Validates and atomically applies state operations to a cloned profile.
///
/// Capability-widening proposals are returned separately and never applied.
pub fn apply_delta(
    profile: &AgentProfile,
    delta: &AgentStateDelta,
    options: ApplyOptions<'_>,
) -> Result<DeltaApplication, ApplyError> {
    let metadata = object(profile.get("metadata"));
    let current = metadata
        .get("revision")
        .and_then(Value::as_u64)
        .unwrap_or(1);
    let target = object(delta.get("target"));
    if target.get("name").and_then(Value::as_str) != metadata.get("name").and_then(Value::as_str) {
        return Err(ApplyError::Message(format!(
            "delta targets {:?} but profile is {:?}",
            target.get("name"),
            metadata.get("name")
        )));
    }
    if target.get("revision").and_then(Value::as_u64) != Some(current) {
        return Err(ConflictError(format!(
            "delta targets revision {:?} but profile is at {current}",
            target.get("revision")
        ))
        .into());
    }
    if let Some(pinned) = target.get("digest").and_then(Value::as_str) {
        if profile_digest(profile).map_err(|error| ApplyError::Message(error.to_string()))?
            != pinned
        {
            return Err(ApplyError::Message(
                "target.digest does not match profile".into(),
            ));
        }
    }
    let lifecycle = object(object(profile.get("spec")).get("lifecycle"));
    let writeback = lifecycle
        .get("writeback")
        .and_then(Value::as_str)
        .unwrap_or("propose");
    if writeback == "off" {
        return Err(ApplyError::Message("lifecycle.writeback is 'off'".into()));
    }
    if writeback == "propose" && !options.approved {
        return Err(ApplyError::Message(
            "lifecycle.writeback is 'propose'; explicit approval is required".into(),
        ));
    }
    let mut working = profile.clone();
    working
        .entry("state")
        .or_insert_with(|| Value::Object(Map::new()));
    let mut warnings = vec![];
    let operations = delta
        .get("operations")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    for (index, raw) in operations.iter().enumerate() {
        apply_operation(
            &mut working,
            &raw.as_object().cloned().unwrap_or_default(),
            &mut warnings,
        )
        .map_err(|error| ApplyError::Message(format!("operation {index}: {error}")))?;
    }
    let stamp = options
        .now
        .unwrap_or_else(Utc::now)
        .to_rfc3339_opts(SecondsFormat::Secs, true);
    enforce_retention(&mut working, &mut warnings, &stamp);
    if let Some(metadata) = working.get_mut("metadata").and_then(Value::as_object_mut) {
        metadata.insert("revision".into(), (current + 1).into());
        metadata.insert("updated_at".into(), stamp.clone().into());
    }
    if !operations.is_empty() {
        if let Some(state) = working.get_mut("state").and_then(Value::as_object_mut) {
            let revision = state.get("revision").and_then(Value::as_u64).unwrap_or(0) + 1;
            state.insert("revision".into(), revision.into());
            state.insert("updated_at".into(), stamp.clone().into());
        }
    }
    let session = object(delta.get("session"));
    let actor = options.actor.unwrap_or("oap-rust");
    let by = session.get("id").and_then(Value::as_str).unwrap_or(actor);
    let mut entry = Map::from_iter([
        ("revision".into(), (current + 1).into()),
        ("at".into(), stamp.clone().into()),
        ("by".into(), by.into()),
        (
            "change".into(),
            delta
                .get("summary")
                .cloned()
                .unwrap_or_else(|| format!("{} state operations", operations.len()).into()),
        ),
        ("sections".into(), json!(["state"])),
    ]);
    if let Some(id) = session.get("id") {
        entry.insert("session_id".into(), id.clone());
    }
    if let Some(harness) = session.get("harness") {
        entry.insert("harness".into(), harness.clone());
    }
    if options.approved {
        entry.insert("approved_by".into(), actor.into());
    }
    working
        .entry("history")
        .or_insert_with(|| Value::Array(vec![]))
        .as_array_mut()
        .unwrap()
        .push(Value::Object(entry));
    enforce_retention(&mut working, &mut warnings, &stamp);
    let pending_proposals = delta
        .get("proposals")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|raw| raw.as_object().cloned())
        .map(|mut proposal| {
            let path = proposal
                .get("path")
                .and_then(Value::as_str)
                .unwrap_or_default();
            if [
                "/spec/tools",
                "/spec/permissions",
                "/spec/memory",
                "/spec/runtime/subagents",
            ]
            .iter()
            .any(|prefix| path.starts_with(prefix))
            {
                proposal.insert("risk".into(), "high".into());
            }
            proposal
        })
        .collect();
    Ok(DeltaApplication {
        profile: working,
        warnings,
        pending_proposals,
    })
}

fn sort_value(entry: &Value, strategy: &str) -> Value {
    let object = object(Some(entry));
    if strategy == "least_confident" {
        object
            .get("confidence")
            .cloned()
            .unwrap_or_else(|| json!(1.0))
    } else if strategy == "oldest" {
        object
            .get("learned_at")
            .or_else(|| object.get("opened_at"))
            .cloned()
            .unwrap_or_default()
    } else {
        ["last_used_at", "updated_at", "learned_at"]
            .iter()
            .find_map(|key| object.get(*key))
            .cloned()
            .unwrap_or_default()
    }
}
fn compare(left: &Value, right: &Value) -> Ordering {
    match (left.as_f64(), right.as_f64()) {
        (Some(a), Some(b)) => a.partial_cmp(&b).unwrap_or(Ordering::Equal),
        _ => left
            .as_str()
            .unwrap_or_default()
            .cmp(right.as_str().unwrap_or_default()),
    }
}

fn enforce_retention(profile: &mut AgentProfile, warnings: &mut Vec<String>, now: &str) {
    let retention =
        object(object(object(profile.get("spec")).get("lifecycle")).get("retention")).clone();
    let strategy = retention
        .get("eviction")
        .and_then(Value::as_str)
        .unwrap_or("least_recently_used");
    if let Some(state) = profile.get_mut("state").and_then(Value::as_object_mut) {
        for collection in ["facts", "preferences"] {
            let Some(entries) = state.get_mut(collection).and_then(Value::as_array_mut) else {
                continue;
            };
            entries.retain(|raw| {
                let entry = object(Some(raw));
                let expired = retention.contains_key("fact_ttl_days")
                    && entry
                        .get("expires_at")
                        .and_then(Value::as_str)
                        .is_some_and(|expiry| expiry < now)
                    && entry.get("pinned") != Some(&Value::Bool(true));
                if expired {
                    warnings.push(format!(
                        "evicted expired {collection} entry {:?}",
                        entry.get("id").and_then(Value::as_str).unwrap_or_default()
                    ));
                }
                !expired
            });
            if collection == "facts" {
                if let Some(cap) = retention
                    .get("max_facts")
                    .and_then(Value::as_u64)
                    .map(|value| value as usize)
                {
                    if entries.len() > cap {
                        let mut candidates: Vec<_> = entries
                            .iter()
                            .filter(|entry| {
                                object(Some(entry)).get("pinned") != Some(&Value::Bool(true))
                            })
                            .cloned()
                            .collect();
                        candidates.sort_by(|a, b| {
                            compare(&sort_value(a, strategy), &sort_value(b, strategy))
                        });
                        let drop_count = entries.len().saturating_sub(cap).min(candidates.len());
                        let dropped: BTreeIds = candidates
                            .into_iter()
                            .take(drop_count)
                            .filter_map(|entry| {
                                object(Some(&entry))
                                    .get("id")
                                    .and_then(Value::as_str)
                                    .map(str::to_owned)
                            })
                            .collect();
                        for id in &dropped {
                            warnings
                                .push(format!("evicted {collection} entry {id:?} ({strategy})"));
                        }
                        entries.retain(|entry| {
                            !object(Some(entry))
                                .get("id")
                                .and_then(Value::as_str)
                                .is_some_and(|id| dropped.contains(id))
                        });
                    }
                }
            }
        }
        if let Some(cap) = retention
            .get("max_open_threads")
            .and_then(Value::as_u64)
            .map(|value| value as usize)
        {
            if let Some(threads) = state.get_mut("open_threads").and_then(Value::as_array_mut) {
                if threads.len() > cap {
                    let mut closed: Vec<_> = threads
                        .iter()
                        .filter(|thread| {
                            matches!(
                                object(Some(thread)).get("status").and_then(Value::as_str),
                                Some("done" | "abandoned")
                            )
                        })
                        .cloned()
                        .collect();
                    closed.sort_by_key(|thread| {
                        object(Some(thread))
                            .get("updated_at")
                            .and_then(Value::as_str)
                            .unwrap_or_default()
                            .to_owned()
                    });
                    let dropped: BTreeIds = closed
                        .into_iter()
                        .take((threads.len() - cap).min(threads.len()))
                        .filter_map(|entry| {
                            object(Some(&entry))
                                .get("id")
                                .and_then(Value::as_str)
                                .map(str::to_owned)
                        })
                        .collect();
                    for id in &dropped {
                        warnings.push(format!("evicted closed thread {id:?}"));
                    }
                    threads.retain(|entry| {
                        !object(Some(entry))
                            .get("id")
                            .and_then(Value::as_str)
                            .is_some_and(|id| dropped.contains(id))
                    });
                    if threads.len() > cap {
                        let remove = threads.len() - cap;
                        threads.drain(..remove);
                        warnings.push(
                            "open_threads still over cap after evicting closed threads".into(),
                        );
                    }
                }
            }
        }
    }
    let history_cap = retention
        .get("max_history")
        .and_then(Value::as_u64)
        .unwrap_or(50) as usize;
    if let Some(history) = profile.get_mut("history").and_then(Value::as_array_mut) {
        if history.len() > history_cap {
            let remove = history.len() - history_cap;
            history.drain(..remove);
        }
    }
}

type BTreeIds = std::collections::BTreeSet<String>;

/// Serializes an OAP document as YAML, JSON, or Markdown.
pub fn serialize(document: &Document, format: crate::OapFormat) -> Result<String, ApplyError> {
    match format {
        crate::OapFormat::Json => serde_json::to_string_pretty(document)
            .map(|text| text + "\n")
            .map_err(|error| ApplyError::Message(error.to_string())),
        crate::OapFormat::Yaml => serde_yaml_ng::to_string(document)
            .map_err(|error| ApplyError::Message(error.to_string())),
        crate::OapFormat::Markdown => {
            let mut copy = document.clone();
            let instructions = copy
                .get_mut("spec")
                .and_then(Value::as_object_mut)
                .and_then(|spec| spec.get_mut("role"))
                .and_then(Value::as_object_mut)
                .and_then(|role| role.remove("instructions"))
                .and_then(|value| value.as_str().map(str::to_owned))
                .unwrap_or_default();
            let yaml = serde_yaml_ng::to_string(&copy)
                .map_err(|error| ApplyError::Message(error.to_string()))?;
            Ok(format!(
                "---\n{}---\n{}\n",
                yaml.trim_start_matches("---\n"),
                instructions.trim_end()
            ))
        }
    }
}

/// Replaces a file atomically using a temporary file in the same directory.
pub fn write_atomically(path: impl AsRef<Path>, data: &[u8]) -> Result<(), ApplyError> {
    let path = path.as_ref();
    let directory = path.parent().unwrap_or_else(|| Path::new("."));
    let temporary = directory.join(format!(
        ".{}.{}.tmp",
        path.file_name()
            .and_then(|name| name.to_str())
            .unwrap_or("oap"),
        std::process::id()
    ));
    let result = (|| {
        let mut file = OpenOptions::new()
            .create_new(true)
            .write(true)
            .open(&temporary)?;
        file.write_all(data)?;
        file.sync_all()?;
        drop(file);
        fs::rename(&temporary, path)?;
        if let Ok(directory) = fs::File::open(directory) {
            directory.sync_all()?;
        }
        Ok::<(), std::io::Error>(())
    })();
    if result.is_err() {
        let _ = fs::remove_file(&temporary);
    }
    result.map_err(|error| ApplyError::Message(error.to_string()))
}
